### PR TITLE
Test debt

### LINKED ISSUE #5326: Extend block adapter unit tests
Add more coverage to the block adapters unit tests, including:

- [x] -  UploadCopyPart
- [x] - UploadCopyPartRange
- [x] - AbortMultiPartUpload
- [x] - GetPreSignedURL
- [ ] - Implement TODOs
- [ ] - Create struct and return as param for extractParamsFromObj https://github.com/treeverse/lakeFS/pull/5317#pullrequestreview-1343798497

### PR DESCRIPTION
**Pull request by nadavsteindler**
Closes #5326 

## Change Description
Closing some unit test debt for the Storage Adapters

### Background

Share context and relevant information for the PR: offline discussions, considerations, design decisions etc.

### Bug Fix

If this PR is a bug fix, please let us know about:

1. Problem - The reason for opening the bug
2. Root cause - Discovered root cause after investigation
3. Solution - How the bug was fixed
      
### New Feature

If this PR introduces a new feature, describe it here.

### Testing Details

How were the changes tested?

### Breaking Change?

Does this change break any existing functionality? (API, CLI, Clients)

## Additional info

Logs, outputs, screenshots of changes if applicable (CLI / GUI changes)

### Contact Details

How can we get in touch with you if we need more info? (ex. email@example.com)

### CODE REVIEW

File: pkg/block/blocktest/adapter.go [hunk 1]
```diff
@@ -463,3 +347,48 @@ func testAdapterWalker(t *testing.T, adapter block.Adapter, storageNamespace str
...
+	preSignedURL, exp, err := adapter.GetPreSignedURL(ctx, obj, block.PreSignModeRead)
+	if adapter.BlockstoreType() == block.BlockstoreTypeGS {
+		require.ErrorContains(t, err, "no credentials found")
+		return
+	} else if adapter.BlockstoreType() == block.BlockstoreTypeLocal {
+		require.ErrorContains(t, err, "operation not supported")
+		return
+	}
+	require.NoError(t, err)
+	require.LessOrEqual(t, exp.Sub(expectedExpiry).Seconds(), 1.0)
```
Comment (eladlachmi) [inline pkg/block/blocktest/adapter.go:366]: I don't love that this test has an inherent race condition. I realize that one second is a lot relative to the expected time scale, but it has a failure mode that does not depend on the test itself or the unit under test. GitHub runners can sometimes be relatively slow.

I can't think of a way to mitigate this completely. However, moving the declaration of `expectedExpiry` to after `GetPreSignedURL` will likely minimize this risk well enough.

**Review by eladlachmi — COMMENTED** [review]
Two comments. Aside from that, looks great!

File: pkg/block/blocktest/multipart.go [hunk 2]
```diff
@@ -0,0 +1,315 @@
...
+	uploadMultiPart(t, ctx, adapter, obj, parts)
+
+	// Begin Multipart Copy
+	resp, err := adapter.CreateMultiPartUpload(ctx, objCopy, nil, block.CreateMultiPartUploadOpts{})
+	require.NoError(t, err)
+	multiParts, err := copyPart(t, ctx, adapter, obj, objCopy, resp.UploadID)
+	if adapter.BlockstoreType() == block.BlockstoreTypeAzure {
+		require.ErrorContains(t, err, "not implemented")
+		return
+	}
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:147]: Why is Azure required to return not implemented?

File: pkg/block/blocktest/multipart.go [hunk 3]
```diff
@@ -0,0 +1,315 @@
...
+
+	// Begin Multipart Copy
+	resp, err := adapter.CreateMultiPartUpload(ctx, objCopy, nil, block.CreateMultiPartUploadOpts{})
+	require.NoError(t, err)
+	multiParts := copyPartRange(t, ctx, adapter, obj, objCopy, resp.UploadID)
+	if adapter.BlockstoreType() == block.BlockstoreTypeAzure {
+		// not implemented in Azure
+		require.Nil(t, multiParts)
+		return
+	}
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:171]: Same here, why is Azure not implemented?

File: pkg/block/blocktest/multipart.go [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+	obj, _ := objPointers(storageNamespace)
+	uploadID := "foobar"
+
+	err := adapter.AbortMultiPartUpload(ctx, obj, uploadID)
+	if err != nil {
+		if adapter.BlockstoreType() == block.BlockstoreTypeLocal {
+			require.ErrorContains(t, err, "invalid upload id")
+		} else {
+			require.ErrorContains(t, err, "404")
+		}
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:101]: It's not recommended to check the returned error.
Basically, if the adapter was expected to return a specific error I would understand, but in this case we are testing the text returned from the underlying storage

File: pkg/block/blocktest/multipart.go [hunk 5]
```diff
@@ -0,0 +1,315 @@
...
+	if err == nil {
+		require.Equal(t, multipartNumberOfParts, len(listResp.Parts))
+	}
+
+	err = adapter.AbortMultiPartUpload(ctx, obj, uploadID)
+	require.NoError(t, err)
+
+	// verify no parts to list
+	verifyListInvalid(t, ctx, adapter, obj, uploadID)
+
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:120]: Same here, why are we testing this? If we would like to have a test that listing with a unknown uploadID fails I would expect to see this in a `testList` of some kind

File: pkg/block/blocktest/multipart.go [hunk 6]
```diff
@@ -0,0 +1,315 @@
...
+		Part: multiParts,
+	})
+	if err != nil {
+		if adapter.BlockstoreType() == block.BlockstoreTypeGS {
+			require.ErrorContains(t, err, "part")
+		} else {
+			require.ErrorContains(t, err, "404")
+		}
+	}
+}
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:131]: Also here, we are testing the underlying storage response, I think it's enough to test that an error is returned

File: pkg/block/blocktest/multipart.go [hunk 7]
```diff
@@ -0,0 +1,315 @@
...
+		require.Equal(t, multipartNumberOfParts, len(listResp.Parts))
+	}
+
+	err = adapter.AbortMultiPartUpload(ctx, obj, uploadID)
+	require.NoError(t, err)
+
+	// verify no parts to list
+	verifyListInvalid(t, ctx, adapter, obj, uploadID)
+
+	_, err = adapter.CompleteMultiPartUpload(ctx, obj, resp.UploadID, &block.MultipartUploadCompletion{
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:121]: What happens if we call AbortMultiPartUpload twice, does that fail, if it is expected to fail, It might be better to test that rather than the completeMultipartUpload here

File: pkg/block/blocktest/multipart.go [hunk 8]
```diff
@@ -0,0 +1,315 @@
...
+
+	multiParts := uploadParts(t, ctx, adapter, obj, resp.UploadID, parts)
+
+	_, err = adapter.CompleteMultiPartUpload(ctx, obj, resp.UploadID, &block.MultipartUploadCompletion{
+		Part: multiParts,
+	})
+	require.NoError(t, err)
+}
+
+// List parts on non-existing part
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:204]: what is a non-existing part? 
If these tests list parts on a non-existing uploadID why is it tested in every test?

File: pkg/block/blocktest/multipart.go [hunk 9]
```diff
@@ -0,0 +1,315 @@
...
+
+func createMultipartContents() ([][]byte, []byte) {
+	parts := make([][]byte, multipartNumberOfParts)
+	var partsConcat []byte
+	for i := 0; i < multipartNumberOfParts; i++ {
+		parts[i] = randstr.Bytes(multipartPartSize + i)
+		partsConcat = append(partsConcat, parts[i]...)
+	}
+	return parts, partsConcat
+}
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:189]: It would be nice to validate somewhere that the last part can be smaller than 5MB

File: pkg/block/blocktest/multipart.go [hunk 10]
```diff
@@ -0,0 +1,315 @@
...
+
+	return multiParts, nil
+}
+
+func getAndCheckContents(t *testing.T, ctx context.Context, adapter block.Adapter, exp []byte, obj block.ObjectPointer) {
+	t.Helper()
+	// first check exists
+	ok, err := adapter.Exists(ctx, obj)
+	require.NoError(t, err)
+	require.Equal(t, true, ok)
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:272]:
```suggestion
	require.True(t, ok)
```

File: pkg/block/blocktest/multipart.go [hunk 11]
```diff
@@ -0,0 +1,315 @@
...
+		return nil, err
+	}
+	multiParts[0].PartNumber = partNumber
+	multiParts[0].ETag = partResp.ETag
+
+	return multiParts, nil
+}
+
+func getAndCheckContents(t *testing.T, ctx context.Context, adapter block.Adapter, exp []byte, obj block.ObjectPointer) {
+	t.Helper()
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:268]: Please document or change name

File: pkg/block/blocktest/multipart.go [hunk 12]
```diff
@@ -0,0 +1,315 @@
...
+
+	reader, err := adapter.Get(ctx, obj)
+	require.NoError(t, err)
+	got, err := io.ReadAll(reader)
+	require.NoError(t, err)
+	requireEqualBigByteSlice(t, exp, got)
+}
+
+// compare two big bytearrays one slice at a time(so that we don't blow up the console on error)
+func requireEqualBigByteSlice(t *testing.T, exp, actual []byte) {
```
Comment (guy-har) [inline pkg/block/blocktest/multipart.go:282]: Why is this needed? 
What's bad with the require.Equal?

File: pkg/block/blocktest/adapter.go [hunk 13]
```diff
@@ -463,3 +342,62 @@ func testAdapterWalker(t *testing.T, adapter block.Adapter, storageNamespace str
...
+	tree := make([]string, 0)
+
+	uri, err := url.Parse(qk.Format())
+	require.NoError(t, err)
+
+	w, err := adapter.GetWalker(uri)
+	require.NoError(t, err)
+
+	walker := store.NewWrapper(w, uri)
+	require.NoError(t, err)
```
Comment (eladlachmi) [inline pkg/block/blocktest/adapter.go:387]: There's no new value for `err` here. No need for this `require.NoError`.

File: pkg/block/blocktest/adapter.go [hunk 14]
```diff
@@ -463,3 +342,62 @@ func testAdapterWalker(t *testing.T, adapter block.Adapter, storageNamespace str
...
+
+	err = walker.Walk(ctx, block.WalkOptions{}, func(e block.ObjectStoreEntry) error {
+		_, p, _ := strings.Cut(e.Address, uri.String())
+		tree = append(tree, p)
+		return nil
+	})
+	if err != nil {
+		t.Fatalf("walking on '%s': %s", uri.String(), err)
```
Comment (eladlachmi) [inline pkg/block/blocktest/adapter.go:395]: This is a bit of an issue. On the one hand, calling `t.Helper()` in this function means the failing test result will point you to the call site - line 167. This is usually what you want from a helper function, especially if it's used in multiple tests/test cases.

However, since this function contains multiple assertions, it could be difficult to tell which one failed.  
`require.NoError` has a third argument, allowing you to add a custom message. The message will be part of the reported test output. This can help you pinpoint the exact place where the error occurred.

File: pkg/block/blocktest/multipart.go [hunk 15]
```diff
@@ -0,0 +1,302 @@
...
+	obj, objCopy := objPointers(storageNamespace)
+	uploadMultiPart(t, ctx, adapter, obj, parts)
+
+	// Begin Multipart Copy
+	resp, err := adapter.CreateMultiPartUpload(ctx, objCopy, nil, block.CreateMultiPartUploadOpts{})
+	require.NoError(t, err)
+	multiParts := copyPartRange(t, ctx, adapter, obj, objCopy, resp.UploadID)
+	if adapter.BlockstoreType() == block.BlockstoreTypeAzure {
+		// not implemented in Azure
+		require.Nil(t, multiParts) // azurite block store emulator did not yet implement this
```
Comment (eladlachmi) [inline pkg/block/blocktest/multipart.go:151]: Why do we need to verify that it's`nil`? If this test doesn't do anything when the block adapter type is Azure, why invest the time in doing any of it? I'd move this check to the top of the test and skip it (return immediately.)

This will always be `nil` until the day that someone happens to upgrade this package (probably while fixing a security issue or upgrading packages across the board for a new Go version or something). Now, they have to deal with this random failing test, which they have no context for and didn't plan for either.  
I can see merit in both approaches and the comments are helpful in this regard, so I'll leave this to your discretion. I'm fine either way.

File: pkg/block/blocktest/multipart.go [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+
+	return multiParts, nil
+}
+
+// getAndCheckContents GETs the Object from the Store, then compares its contents to the exp byte array
+func getAndCheckContents(t *testing.T, ctx context.Context, adapter block.Adapter, exp []byte, obj block.ObjectPointer) {
+	t.Helper()
+	// first check exists
+	ok, err := adapter.Exists(ctx, obj)
+	require.NoError(t, err)
```
Comment (eladlachmi) [inline pkg/block/blocktest/multipart.go:258]: I'd also add custom messages here (like I explained above), as this can fail in multiple assertions.

**Review by eladlachmi — CHANGES_REQUESTED** [review]
Looking good! I still think there are a few things we can improve on.

**Review by eladlachmi — APPROVED** [review]
LGTM! Well done 🎉
